consensus/propagation/types/types_test.go

```diff
@@ -9,8 +10,146 @@ import (
 	"github.com/tendermint/tendermint/crypto/tmhash"
 	"github.com/tendermint/tendermint/libs/bits"
 	"github.com/tendermint/tendermint/libs/rand"
+	"github.com/tendermint/tendermint/types"
 )
 
+func TestTxMetaData_RoundTrip(t *testing.T) {
+	tests := []struct {
+		name string
+		data *TxMetaData
+	}{
+		{
+			"valid metadata",
+			&TxMetaData{Hash: rand.Bytes(tmhash.Size), Start: 0, End: 10},
+		},
+		{
+			"empty hash",
+			&TxMetaData{Hash: []byte{}, Start: 5, End: 10},
+		},
+		{
+			"same start and end",
+			&TxMetaData{Hash: rand.Bytes(tmhash.Size), Start: 7, End: 7},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			proto := tt.data.ToProto()
+			roundTripped := TxMetaDataFromProto(proto)
+			require.Equal(t, tt.data, roundTripped)
+		})
+	}
+}
+
+func TestTxMetaData_ValidateBasic(t *testing.T) {
+	tests := []struct {
+		name    string
+		data    *TxMetaData
+		expects error
+	}{
+		{
+			"valid metadata",
+			&TxMetaData{Hash: rand.Bytes(tmhash.Size), Start: 0, End: 10},
+			nil,
+		},
+		{
+			"invalid start > end",
+			&TxMetaData{Hash: rand.Bytes(tmhash.Size), Start: 10, End: 5},
+			errors.New("TxMetaData: Start > End"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.data.ValidateBasic()
+			if tt.expects != nil {
+				assert.EqualError(t, err, tt.expects.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestCompactBlock_RoundTrip(t *testing.T) {
+	mockProposal := types.NewProposal(
+		4, 2, 2,
+		makeBlockID(rand.Bytes(tmhash.Size), 777, rand.Bytes(tmhash.Size)),
+	)
+	mockProposal.Signature = rand.Bytes(types.MaxSignatureSize)
+	tests := []struct {
+		name string
+		data *CompactBlock
+	}{
+		{
+			"valid block",
+			&CompactBlock{
+				BpHash:    rand.Bytes(tmhash.Size),
+				Blobs:     []*TxMetaData{{Hash: rand.Bytes(tmhash.Size), Start: 0, End: 10}},
+				Signature: rand.Bytes(types.MaxSignatureSize),
+				Proposal:  *mockProposal,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			proto := tt.data.ToProto()
+			// Assuming CompactBlockFromProto exists
+			roundTripped, err := CompactBlockFromProto(proto)
```

Comment:
this is a serialisation/deserialisation test no? why is it called round trip?

 By round trip, I understand it's a test that tests the whole workflow of a compact block

Reply:
we're "round tripping" the encoding and decoding